### DISCUSSION TITLE
New Lanbon L8 model: AC fan. Please add

### DISCUSSION THREAD

**Discussion by samturner3** [HASwitchPlate/openHASP#352] [Hardware]
There is a new Lanbon L8 model which has output for an AC fan.
This is the listing (white labeled as ctec sold for Australia)
https://www.amazon.com.au/gp/product/B0B4VWHTJD/

_I have the normal 3 gang ctec and flashed successfully with openHASP and IS for all intents and purposes a Lanbon L8._

There is 2 physical screw terminal outputs, one for light (normal relay) and one for fan which I assume just switches between 3 capacitor circuits (including 0/no capacitor for full speed).

I have not yet confirmed what GPIO pins activate what circuit, however I suspect it is similar logic to this [ESP home device logic](https://www.esphome-devices.com/devices/Deta-Grid-Connect-Smart-Fan-Speed-Controller-with-Touch-Light-Switch) where we have 2 or 3 gpios, and depending on the state the power will be routed through different capacitor circuits.

Would be good to get this type included with openHASP.
I have a unit if you need info, once I wire it to the fan I can let you know which gpios need to be switched for the 3 speeds.
I may even hold off flashing openHASP and test what GPIOs are active for the speeds using the stock firmware.

**Comment by fvanroie** [comment]
That's a great find. Some questions though:
- Does it have a model on the box? like `L8-HD`, `L8-HS` or `L8-HT`
- What's the PCB version of the plate?

The GPIO assignment will be needed to add support for this device. I suspect `K1-K3`might be the speed indicators, with an additional pin for light. However, they could also use 2 pins for indicating a binary speed `0-3`... Powering the faceplate with 5V while adding a resistor with a LED to `K1-K3´ pins should give you the assignments using the original firmware.

**Reply by peter-lill** [reply]
Hey guys,

Just following up on this one as couldn't see it had progressed.

I now have one of these, have a functioning light but no fan as yet!

Gpio pin 27 is the light relay

http://www.ctec.com.au/the-mirror-range outlines the models, L8-HFU being the fan

**Reply by fvanroie** [reply]
Do you still have the original firmware for testing? It seems the L8-HFU is only available on the Australian market...
You'll need to check the pin assignments with a multi-meter to see which pins are connected to which functionality.
If you could test powering the faceplate with 5V while adding a resistor with a LED to `K1-K3´ pins that would be great.
That should give you the assignments using the original firmware.

Please provide a picture of the PCB and board version.

**Comment by peter-lill** [comment]
Unfortunately do not have the original firmware.

Have purchased another unit so will check this out next week

**Comment by james-t3l-co** [comment]
Did anyone get the GPIO details for the L8-HFU? I'm guessing by the few responses over the years and nothing recent not many people seem to have bought these?

I've got 2 units that I would love to flash with openhasp 0.6 (still running that in the rest of the house) but without knowing the GPIO settings, won't help of course.

The board is labelled HS_CB_V1.17(20210805).

I've grabbed a copy of the stock firmware (took the whole 8mb dump) and can power it using 5v and test pins, but I am not sure what I am looking for.

Based on Peter-Lill's comment, GPIO 27 as a relay seems to control the light and that seems to ring true based on this extract of the minicom output at boot:

DEVICE_DRIVER_METHOD_BY_SLAVE_MCU
I (169) gpio: GPIO[27]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (383) gpio: GPIO[22]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
fans exRelay set:1.  <- pressed light button on
fans exRelay set:0. <- pressed light button off
fans exRelay set:0. <- pressed fan power 1 (with light still off)
fans exRelay set:0. <- pressed fan power 2 (with light still off)
fans exRelay set:0. <- pressed fan power 3 (with light still off)
fans exRelay set:0. <- pressed fan power off (with light still off)
fans exRelay set:1.  <- pressed light button on
fans exRelay set:1. <- pressed fan power 1 (with light still on)

So nothing is output via serial interface when pressing the fan buttons, except to indicate what the light switch relay is set to.

There are no relay clicks from what I can tell when switching fan speeds (no clicking on the device) as opposed turning the light on and off where I hear a click.

Checking voltage on the K1-K3 isn't obvious as to whats happening - i'm using a cheap multimeter though.
K3 remains at 3.2v no matter what light or fan button is engaged (off, 1, 2 or 3) on the fan.
K2 remains at 0v no matter what light or fan button is engaged on the display.
K1 seems to be related to the light switch - when the light is on via the display, it's 0v - when its off it returns to 1.6v

I don't have any 5v leds with resistors to test but can get some if someone can point me in the right direction as to what i would be looking for with these (there was an earlier mention in this thread about using a 5v led with resistor to map out what K1-K3 do.

I'm guessing the fan is always in the on position and the gearing is set to the capacitor (off, 30%, 60%, 100% etc).
Any suggestions on figuring out what does what on this device? I am guessing GPIO22 is just the display stuff?

**Reply by james-t3l-co** [reply]
OK, this took me a week to find time to figure it all out.

K1 seems to be the light switch GPIO (K1 active low)?
K3 is being used as output for UART using a 4 byte command to control the fan (I was not getting decent readings off the multimeter which led me to believe it was being used to transmit UART. Connecting a TTYUSB device ground to ground and RX to K3 on the breakout pins and checking output when pressing buttons was able to capture the UART output. 

- UART Packets on K3 are used to control the fan speed (EF 02 XX YY) - XX being the speed and YY being checksum) 
- Seems to have a secondary power MCU controlling Fan output gang - aligns with the boot statement DEVICE_DRIVER_METHOD_BY_SLAVE_MCU
- Likely Phase angle triac control (as it lets me set a % - can also be set in the UI) rather than having a fixed stepping pattern etc

Anyone able to give me any insights of how I can get this working in Openhasp now? I'd really like to do away with the terrible UI that won't properly connect to their app etc.

4 byte protocol seems to be:
Byte 1: 0xEF
Byte 2: 0x02
Byte 3: Percentage (0 - 100)
Byte 4: XOR shecksum

K3 has continuity to the 14th PIN on the ESP32-WROVER-B chip (IO12 I think). It's the PIN above the GND2 pin.

**Reply by fvanroie** [reply]
It seems that this is the same [protocol as for the L8-HD dimmer](https://github.com/HASwitchPlate/openHASP/discussions/380).

https://community.home-assistant.io/t/interesting-touch-switch-lanbon-l8/264112/39
You should be able to use it in [the same way](https://openhasp.com/0.7.0/firmware/configuration/gpio/#type_1)...

**Reply by james-t3l-co** [reply]
Thanks! This will make things so much easier, love your work!!

I'll need to fumble my way through the code perhaps to add a L8-HFU model somehow. If I select L8-HD(AU or EU) on the GPIO pin assignment (using 0.6.3) it doesn't seem to get the whining noise from the unit which happens with stock firmware.

When the device is booted, there seems to be an initialisation command issued to the MCU on IO12 and it might be a different code.

It's issuing EF 01 4E A0 as the intialisation code (i presume its an initialisation code as it only happens at boot and is EF 01 prefix) at boot before immediately sending EF 02 00 ED (setting fan speed to 0%).

Edit:
Looking at hasp_gpio.cpp the other dimmers seem to use a different code for initialisation:
 const char command[9] = "\xEF\x01\x4D\xA3"; // Start Lanbon Dimmer

So I presume I would need to set a new hasp_gpio_type in a bunch of places and call it something like SERIAL_DIMMER_FAN and in the case select for that use
            const char command[9] = "\xEF\x01\x4E\xA0"; // Start Lanbon Fan Dimmer

~~Also, the values for this device seem to be 0 to 100 rather than 0 to 255? IE when i hit the 100% on the original firmware it issues a EF 02 64 89 where 0x64 is 100 rather than 255 which I think the light dimmer sets?~~ Ignore that, sorry I think I had that wrong and the serial commands are being sent from 0 to 100 just converted from 0 to 255 for slider events or something like that.

**Reply by fvanroie** [reply]
Try changing the init code first to see if it actually works for your device.
Then we can see how to integrate it.

**Comment by james-t3l-co** [comment]
I put in a new device called L8-HSU (AU) - used the initialisation code I grabbed out of the stock firmware for it; set a max of 100 for the value for brightness and it's all working through home assistant (not the prettiest but certainly working).

In HA, as it's getting a brightness tag in MQTT - the slider is 0-255 but it's got a max of 100 so at full speed it's pulled 1/3 way across the slider, but it does the trick for the moment.

<img width="675" height="535" alt="image" src="https://github.com/user-attachments/assets/6f9f0ec8-23da-48b9-8c75-ec2128e47dab" />

<img width="633" height="394" alt="image" src="https://github.com/user-attachments/assets/aad0c657-4324-4160-8ac9-505f20ca837a" />

So it's certainly working. Fan is working as expected - just mislabelled in HA using MQTT autodiscovery etc but will look at that a bit later, wanted functionality so i can set the fan from the other room and I've got that now :)

**Comment by james-t3l-co** [comment]
Well, this device is doing my head in :).
This was working well for a few days (could set the 3 speeds via HA).

Now, I can't get the dimmer to work under openhasp builds. I've tried the builds it was working on above; it will either be on (high) or off now - no speed control (low/medium).

If I switch back to the stock firmware on same panel/fan/wall socket -  the device works, can set off, 1, 2 or 3. I'm not sure what's going on to be honest.

I've checked the correct signals are sent using mincom attached to K3, both devices (at least when not connected to the wall socket) are outputting the same stuff (initialisation code) - then setting appropriate speeds (i'm using 0%, 30%, 60% or 100%) in openhasp to match the codes output by the stock firmware.

I'm not sure what to make of it. When I plug in the panel with a stock firmware, I hear what I presume is the dimmer module making high pitched noise in the background at the wall socket, but using an openhasp build; it's not making any noise.

I am thinking it is sending something else to start the dimmer module via K3 (potentially if K2 is the return serial pin) but can't get that without attaching pins to the live unit and not particularly keen on doing that.

Might need to poke around the original firmware and see if i can get anything out of that.
